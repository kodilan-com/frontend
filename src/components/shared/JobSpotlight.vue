<script>
import { mapState, mapActions } from 'vuex';
import Loader from './Loader';
import JobTypeBadge from './JobTypeBadge';

export default {
  props: {
    contentLength: {
      type: Number,
      required: false,
      default: 200,
    },
  },
  components: {
    JobTypeBadge,
    Loader,
  },
  data() {
    return {
      isLoading: true,
    };
  },
  computed: {
    ...mapState(['featuredPosts']),
  },
  methods: {
    ...mapActions(['fetchFeaturedPosts']),
    getPostSummary(description) {
      const tmp = document.createElement('div');
      tmp.innerHTML = description;
      const descText = tmp.textContent || tmp.innerText || '';
      const cl = this.contentLength;

      return descText.length > cl ? `${descText.slice(0, cl)}...` : descText;
    },
  },
  created() {
    this.fetchFeaturedPosts()
      .then(() => {
        this.isLoading = false;
      });
  },
};
</script>

<template>
  <div class="columns">
    <h3>
      <svg xmlns="http://www.w3.org/2000/svg" fill="none" viewBox="0 0 24 24" stroke="currentColor">
        <path stroke-linecap="round" stroke-linejoin="round" stroke-width="1.5" d="M13 7h8m0 0v8m0-8l-8 8-4-4-6 6" />
      </svg>
      Öne çıkan ilanlar
    </h3>
    <loader v-if="isLoading" />
    <template v-else>
      <div id="job-spotlight">
        <div class="showbiz">
          <div class="wrapper">
            <ul>
              <li v-for="post in featuredPosts" :key="post.slug">
                <div class="job-spotlight">
                  <router-link :to="`/ilanlar/${post.slug}`" class="job-title">
                    {{ post.position }}
                  </router-link>
                  <div class="details">
                    <span>
                      <i class="fa fa-briefcase" />
                      <router-link :to="`/firmalar/${post.company.slug}`">
                        {{ post.company.name }}
                      </router-link>
                    </span>
                    <span>
                      <i class="fa fa-map-marker" />
                      <router-link :to="`/ilan-ara?location=${post.location}`">
                        {{ post.location }}
                      </router-link>
                    </span>
                    <job-type-badge :post="post" />
                    <p>{{ getPostSummary(post.description) }}</p>
                    <div v-for="tag in post.tags" :key="tag.slug" class="listing-date post-tag">
                      <router-link :to="`/etiket/${tag.slug}`">
                        {{ tag.name }}
                      </router-link>
                    </div>
                    <router-link :to="`/ilanlar/${post.slug}`" class="button tag-go-featured-post">
                      İlana git &nbsp;&nbsp;<i class="fa fa-angle-right" />
                    </router-link>
                  </div>
                </div>
              </li>
            </ul>
            <div class="clearfix" />
          </div>
        </div>
      </div>
    </template>
  </div>
</template>

<style lang="scss">

  h3 {
    display: flex;
    flex-direction: row;
    align-items: center;

    font-weight: 500;
    margin-bottom: 8px;

    svg {
      width: 24px;
      height: 24px;
      background-color: #26ae61;
      stroke: #FFF;
      margin-right: 12px;
      padding: 8px;
      border-radius: 50%;
    }
  }


#job-spotlight {
  .wrapper {
    overflow: hidden;

    ul {
      overflow: hidden;
      position: relative;
      transition: left 0.3s ease-in;
    }

    li {
      float: left;
      width: 100%;
    }
  }

  .post-tag {
    margin: 5px 5px 5px 0;
  }
}
</style>
