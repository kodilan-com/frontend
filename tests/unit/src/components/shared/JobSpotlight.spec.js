import { shallowMount, createLocalVue, RouterLinkStub } from '@vue/test-utils';
import JobSpotlight from '@/components/shared/JobSpotlight';
import Vuex from 'vuex';

const localVue = createLocalVue();

localVue.use(Vuex);

describe('JobSpotlight.vue', () => {
  let store;
  let jobSpotlightStore;
  let wrapper;
  const renderComponent = (options) => {
    wrapper = shallowMount(JobSpotlight, {
      stubs: {
        RouterLink: RouterLinkStub,
      },
      localVue,
      store,
      ...options,
    });
  };
  const getHrefs = selector => wrapper.findAll(selector).wrappers.map(job => job.props('to'));

  beforeEach(() => {
    jobSpotlightStore = {
      state: {
        featuredPosts: [{
          slug: 'slug-1',
          location: 'postLocation1',
          description: 'Post Description 1',
          company: {
            name: 'Company Name 1',
            slug: 'company-slug-1',
          },
          tags: [{
            slug: 'post-1-tag-1',
            name: 'Post 1 Tag 1 Name',
          }, {
            slug: 'post-1-tag-2',
            name: 'Post 1 Tag 2 Name',
          }],
        }],
      },
      actions: {
        fetchFeaturedPosts: jest.fn(),
      },
    };
    store = new Vuex.Store(jobSpotlightStore);

    renderComponent();
  });

  afterEach(() => {
    wrapper.destroy();
  });

  it('should show loader when fetching featured posts is done', (done) => {
    jest.useFakeTimers();
    jest.advanceTimersByTime(100);

    jobSpotlightStore.actions.fetchFeaturedPosts.mockResolvedValue(true);

    renderComponent();

    expect(jobSpotlightStore.actions.fetchFeaturedPosts).toBeCalled();
    expect(wrapper.find('loader-stub').exists()).toBe(true);
    jest.useRealTimers();


    setTimeout(() => {
      expect(wrapper.find('loader-stub').exists()).toBe(false);

      done();
    }, 101);
  });

  it('should match snapshot', () => {
    expect(wrapper.element).toMatchSnapshot();
  });

  it('should cuts the description and put three dots on the end if description length is greater then 200', async () => {
    jobSpotlightStore.state.featuredPosts[0].description = global.dummy.string(201);

    renderComponent();

    await localVue.nextTick();
    await localVue.nextTick();

    expect(wrapper.find('.details p').text().length).toBe(203);
  });

  it('should pass the post to JobTypeBadge component', () => {
    expect(wrapper.find('job-type-badge-stub').props('post')).toEqual(jobSpotlightStore.state.featuredPosts[0]);
  });

  it('should generate job url correctly', () => {
    expect(getHrefs('.job-title'))
      .toEqual(jobSpotlightStore.state.featuredPosts.map(post => `/ilanlar/${post.slug}`));
    expect(getHrefs('.button.tag-go-featured-post'))
      .toEqual(jobSpotlightStore.state.featuredPosts.map(post => `/ilanlar/${post.slug}`));
  });

  it('should generate company links correctly', () => {
    expect(getHrefs('.fa.fa-briefcase + a'))
      .toEqual(jobSpotlightStore.state.featuredPosts.map(post => `/firmalar/${post.company.slug}`));
  });

  it('should generate post location links correctly', () => {
    expect(getHrefs('.fa.fa-map-marker + a'))
      .toEqual(
        jobSpotlightStore.state.featuredPosts.map(post => `/ilan-ara?location=${post.location}`),
      );
  });

  it('should generate post tags link correctly', () => {
    expect(getHrefs('.listing-date.post-tag a'))
      .toEqual(jobSpotlightStore.state.featuredPosts
        .flatMap(post => post.tags.map(tag => `/etiket/${tag.slug}`)));
  });

  it('should getPostSummary ignore html tags while calculating description length', () => {
    const description = global.dummy.string(200);

    expect(wrapper.vm.getPostSummary(`<div>${description}</div>`))
      .toBe(description);
  });

  it('should getPostSummary return empty string if description is empty', () => {
    expect(wrapper.vm.getPostSummary(''))
      .toBe('');
  });
});
