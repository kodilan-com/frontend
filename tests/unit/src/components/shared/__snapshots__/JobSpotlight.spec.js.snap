// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`JobSpotlight.vue should match snapshot 1`] = `
<div
  class="columns"
>
  <h3
    class="margin-bottom-5"
  >
    
    Öne çıkan ilanlar
  
  </h3>
   
  <div
    id="job-spotlight"
  >
    <div
      class="showbiz"
    >
      <div
        class="wrapper"
      >
        <ul>
          <li>
            <div
              class="job-spotlight"
            >
              <a
                class="job-title"
              >
                
                  
                
              </a>
               
              <div
                class="details"
              >
                <span>
                  <i
                    class="fa fa-briefcase"
                  />
                   
                  <a>
                    
                      Company Name 1
                    
                  </a>
                </span>
                 
                <span>
                  <i
                    class="fa fa-map-marker"
                  />
                   
                  <a>
                    
                      postLocation1
                    
                  </a>
                </span>
                 
                <job-type-badge-stub
                  post="[object Object]"
                />
                 
                <p>
                  Post Description 1
                </p>
                 
                <div
                  class="listing-date post-tag"
                >
                  <a>
                    
                      Post 1 Tag 1 Name
                    
                  </a>
                </div>
                <div
                  class="listing-date post-tag"
                >
                  <a>
                    
                      Post 1 Tag 2 Name
                    
                  </a>
                </div>
                 
                <a
                  class="button tag-go-featured-post"
                >
                  
                    İlana git   
                  <i
                    class="fa fa-angle-right"
                  />
                </a>
              </div>
            </div>
          </li>
        </ul>
         
        <div
          class="clearfix"
        />
      </div>
    </div>
  </div>
</div>
`;
